### PR TITLE
Managed aligned memory

### PR DESCRIPTION
**Pull request by kimishpatel**
1. Managed DLTensor when input PT Tensor is unaligned.
       It maintains allocation and eallocation DLTensors corresponding
       PT Values and maintains a map of it.
    2. Managed param tensor just wraps that and assumes that parameters
       are immutable. Thus if a param tensor is created using
       ManagedParamTensors then it will return a cached copy of the tensor
       corresponding to the Value* if one exists. If not allocation and copy
       from corresponding PT tensor is performed. This happens only if
       input PT tensor is unaligned.
    3. Adds a way to annotate ops to specify which of their inputs are
       parameters that are expected to be immutable.

Need to add a few test to validate no memory leaks.

### CODE REVIEW

**Comment by kimishpatel** [comment]
Depends on this PR: https://github.com/pytorch/tvm/pull/85

File: torch_tvm/operators.h [hunk 1]
```diff
@@ -3,6 +3,12 @@
 #include <tvm/relay/expr.h>
 #include <tvm/relay/op.h>
 
+#define PARAM_INDICES_convolution {1, 2}
+#define PARAM_INDICES_layer_norm {2, 3}
+#define PARAM_INDICES_linear {1, 2}
+
+#define PARAM_INDICES(op_name) PARAM_INDICES_##op_name
```
Comment (bwasti) [inline torch_tvm/operators.h]: this is a bit strange, can we inline the {1, 2} etc into the list of ops as the final param?

File: torch_tvm/memory_utils.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+namespace utils {
+
+struct DLManagedTensorDeleter {
+  void operator()(DLManagedTensor* manager_ctx) {
+    if (manager_ctx == nullptr)
+      return;
+    auto dl_tensor = manager_ctx->dl_tensor;
+    if (dl_tensor.data) {
+      delete dl_tensor.data;
+      // Implicit assumption is made that data, shape and strides are
```
Comment (bwasti) [inline torch_tvm/memory_utils.h]: lets make this explicit with a CHECK

File: torch_tvm/memory_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  auto num_dims = tensor.dim();
+  dl_tensor.ndim = num_dims;
+  dl_tensor.dtype = at::getDLDataType(tensor);
+  int64_t device_id = 0;
+  dl_tensor.ctx = getDLContext(tensor, device_id);
+  dl_tensor.shape = dl_tensor.strides = nullptr;
+  dl_tensor.data = nullptr;
+  dl_tensor.shape = new int64_t[num_dims];
+  dl_tensor.strides = new int64_t[num_dims];
+  AT_CHECK(dl_tensor.shape != nullptr && dl_tensor.strides != nullptr,
```
Comment (bwasti) [inline torch_tvm/memory_utils.cpp]: AT_CHECK is deprecated, use TORCH_CHECK

File: torch_tvm/memory_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#include "memory_utils.h"
+
+namespace torch_tvm {
+namespace utils {
+
+bool is_aligned(void* data_ptr, std::uintptr_t alignment_in_bytes) {
+  auto mask = alignment_in_bytes - 1;
+  AT_CHECK((alignment_in_bytes & mask) == 0);
```
Comment (bwasti) [inline torch_tvm/memory_utils.cpp]: might be nice to make this a compile time check

File: torch_tvm/compiler.cpp [hunk 5]
```diff
@@ -320,8 +342,22 @@ void TVMCompiler::run(Stack& stack) {
...
+    auto tensor = ivalue.toTensor();
+    DLManagedTensor* dl_tensor;
+    if (torch_tvm::utils::is_aligned(tensor.data_ptr(),
+          tvm::runtime::kAllocAlignment)) {
+      dl_tensor = at::toDLPack(tensor);
+    } else if (cache_[spec].tvm_param_tensors_.value_exists(value)) {
+      // Is the value parameter that is assumed to immutable?
+      dl_tensor = cache_[spec].tvm_param_tensors_.
+        get_tensor_like_for_value(value, tensor);
+    } else {
```
Comment (bwasti) [inline torch_tvm/compiler.cpp]: isn't this final case unsound?

File: torch_tvm/compiler.cpp [hunk 6]
```diff
@@ -320,8 +342,22 @@ void TVMCompiler::run(Stack& stack) {
...
-    auto tensor = ivalue.toTensor().to(at::kFloat);
-    auto dl_tensor = at::toDLPack(tensor);
+    //auto tensor = ivalue.toTensor().to(at::kFloat);
+    auto tensor = ivalue.toTensor();
+    DLManagedTensor* dl_tensor;
+    if (torch_tvm::utils::is_aligned(tensor.data_ptr(),
+          tvm::runtime::kAllocAlignment)) {
+      dl_tensor = at::toDLPack(tensor);
+    } else if (cache_[spec].tvm_param_tensors_.value_exists(value)) {
+      // Is the value parameter that is assumed to immutable?
```
Comment (bwasti) [inline torch_tvm/compiler.cpp]: let's chuck in a warning here just in case

File: torch_tvm/compiler.h [hunk 7]
```diff
@@ -10,12 +10,15 @@
...
 
+#include "managed_tensors.h"
+
 struct TVMObject {
   tvm::PackedFunc kernel;
   tvm::PackedFunc set_input;
   tvm::PackedFunc get_output;
   // Map input indices to values in the subgraph
   std::vector<torch::jit::Value*> input_values;
+  torch_tvm::ManagedParamTensors tvm_param_tensors_;
```
Comment (bwasti) [inline torch_tvm/compiler.h]: nit: no trailing `_`, this is a public variable

File: torch_tvm/compiler.cpp [hunk 8]
```diff
@@ -126,6 +129,7 @@ tvm::relay::Expr TVMCompiler::convertToRelay(
 tvm::relay::Function TVMCompiler::convertToRelay(
     std::shared_ptr<Graph> subgraph,
     TVMContext ctx,
+    torch_tvm::ManagedParamTensors* tvm_param_tensors_ptr,
```
Comment (bwasti) [inline torch_tvm/compiler.cpp]: can we just use `std::vector<size_t>* immutable_input_indices`?  Adding a new data-structure should be avoided here

File: torch_tvm/compiler.cpp [hunk 9]
```diff
@@ -148,7 +152,15 @@ tvm::relay::Function TVMCompiler::convertToRelay(
       auto uses = value->uses();
       for (const auto& use : uses) {
         tvm::Array<tvm::relay::Expr> relay_inputs;
+        // Things like prim::Return
+        // Should we be more explicit here?
+        // That only prim::Return should be skipped?
+        if (use.user->outputs().size() < 1) {
```
Comment (bwasti) [inline torch_tvm/compiler.cpp]: what is this logic doing here?

File: torch_tvm/compiler.cpp [hunk 10]
```diff
@@ -169,15 +181,18 @@ tvm::relay::Function TVMCompiler::convertToRelay(
               }
             }
           }
+          if (!skip_user &&
+              tvm_param_tensors_ptr &&
+              std::find(param_indices.begin(),
+              param_indices.end(), input_index) != param_indices.end()) {
+            tvm_param_tensors_ptr->insert_value(input);
```
Comment (bwasti) [inline torch_tvm/compiler.cpp]: this is much easier to understand with a standard container instead of a custom data structure

File: torch_tvm/managed_tensors.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+      auto managed_tensor_ptr = torch_tvm::utils::
+        alloc_and_copy_data(tensor);
+      managed_tensors[val] = DLManagedTensorPtr(managed_tensor_ptr);
+    }
+
+  protected:
+    std::unordered_map<const Value*, DLManagedTensorPtr> managed_tensors;
+};
+
+class ManagedParamTensors : public ManagedTensors {
```
Comment (bwasti) [inline torch_tvm/managed_tensors.h]: why is this needed?

File: torch_tvm/memory_utils.cpp [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  return (reinterpret_cast<std::uintptr_t>(data_ptr) & mask) == 0;
+}
+
+DLManagedTensor* alloc_and_copy_data(const at::Tensor& tensor) {
+  DLManagedTensor* dl_managed_tensor = new DLManagedTensor();
+  // managed_tensor_deleter is supplied to unique_ptr as a deleter
+  // of this managed memory. Thus setting deleter to nullptr;
+  dl_managed_tensor->deleter = nullptr;
```
Comment (bwasti) [inline torch_tvm/memory_utils.cpp]: why are we hijacking the lifetime logic? can we hand over a deleter that works with DLManagedTensor?  we don't want to have to maintain this lifetime stuff

File: torch_tvm/operators.h [hunk 13]
```diff
@@ -11,17 +17,23 @@ tvm::relay::Expr getOperator(
...
 uint64_t getNoneSentinel();
 
+const std::vector<int32_t>& getParamIndices(torch::jit::Node* node);
+
 using TVMOpFunctor = std::function<tvm::relay::Expr(
     torch::jit::Node* node,
     tvm::Array<tvm::relay::Expr> inputs)>;
 using TVMScheduleFunctor = std::function<const tvm::runtime::PackedFunc*()>;
 
+using ParamIndicesType = std::vector<int32_t>;
```
Comment (bwasti) [inline torch_tvm/operators.h]: no need for this, it's readable without it

File: torch_tvm/operators.cpp [hunk 14]
```diff
@@ -216,7 +222,8 @@ RegisterTVMOperator reg({
              bias_add_op, {out, inputs[2]}, tvm::Attrs(bias_add_attrs), {});
        }
        return out;
-     }},
+     },
+     "", PARAM_INDICES(convolution)},
```
Comment (bwasti) [inline torch_tvm/operators.cpp]: re my comment below, this would be fine as
```
}, "", {1, 2}},
```

File: torch_tvm/operators.cpp [hunk 15]
```diff
@@ -23,12 +23,18 @@ std::unordered_map<Symbol, TVMOpFunctor>& getTVMOperatorMap() {
   return map;
 }
 
+std::unordered_map<Symbol, ParamIndicesType>& getOpParamsMap() {
```
Comment (bwasti) [inline torch_tvm/operators.cpp]: I wonder if we should change getTVMOperatorMap to just return the `struct TVMOpMap` (and maybe even rename that to something a bit more understandable)

File: torch_tvm/compiler.cpp (hunk 9 above)
Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: I just moved the logic up here. It was always there. We do not translate return to any tvm op. So we dont translated it.

File: torch_tvm/compiler.cpp (hunk 5 above)
Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: We need this if the input data, not parameter, is unaligned.

File: torch_tvm/managed_tensors.h (hunk 11 above)
Comment (kimishpatel) [reply torch_tvm/managed_tensors.h]: I was under the impression that the deleter had to be explicitly called and given we are creating this tensor who would call the deleter. But it seems NDArray does call the deleter of underlying DLManagedTensor if specified. So yes I agree, we can get rid of much of these wrappers.

Comment (kimishpatel) [reply torch_tvm/managed_tensors.h]: @bwasti ^^

File: torch_tvm/compiler.cpp (hunk 8 above)
Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: Why is that? Note that in the current implementation this is needed. We cannot just get away with the indices.

File: torch_tvm/compiler.cpp (hunk 10 above)
Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: This is inserting the Value pointer in the cache. Perhaps if `tvm_param_tensor_ptr` is names as `tvm_param_cache_ptr` it might be more readable.

File: torch_tvm/memory_utils.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch_tvm/memory_utils.cpp]: Well if we constexpr alignment to 64 or some value then perhaps we can convert this to static_assert. Then rename this to is_64B_aligned. But the alignment requirement comes from tvm so even though we know it needs 64B aligned, making it constexpr here will decouple it from tvm. Unfortunately due to some include issues I had to hardcode that value to 64 in some places, but that I intend to fix.

File: torch_tvm/memory_utils.cpp (hunk 12 above)
Comment (kimishpatel) [reply torch_tvm/memory_utils.cpp]: Read my comment above. I think we will have to still maintain the cache of some sort. If we dont want to manage lifetime we can hand it over to NDArray the cache will require to map Value* to NDArray*. If we agree on that, I think that is doable.

File: torch_tvm/operators.cpp (hunk 15 above)
Comment (kimishpatel) [reply torch_tvm/operators.cpp]: It might be better. I had to add `getOpParamsMap` and corresponding `param_map` because we dont save TVMOpMap. Also probably should rename `TVMOpMap` to `TVMOp`?

File: torch_tvm/operators.cpp (hunk 14 above)
Comment (kimishpatel) [reply torch_tvm/operators.cpp]: So I dont like constant 1s and 2s lying around without explaining what they are. `PARAM_INDICES` conveys clearly what they are. If you think it is overly complicated to go somewhere else and add a define for a new op, I understand that. But note that it is cleaner this way. I would have `op_params.h` where all such defines exist. We can also make that more readable by having comments there.

File: torch_tvm/operators.h (hunk 1 above)
Comment (kimishpatel) [reply torch_tvm/operators.h]: See my comment above as to why I chose to do this way.

File: torch_tvm/managed_tensors.h [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+#include <tvm/runtime/device_api.h>
+
+#include <torch/csrc/jit/ir.h>
+#include <ATen/DLConvertor.h>
+
+#include "memory_utils.h"
+
+namespace torch_tvm {
+namespace {
+}
```
Comment (bwasti) [inline torch_tvm/managed_tensors.h]: rm

**Comment by yinghai** [comment]
Fixes https://github.com/pytorch/tvm/issues/80

File: torch_tvm/memory_utils.cpp (hunk 4 above)
Comment (yinghai) [reply torch_tvm/memory_utils.cpp]: You can make `alignment_in_bytes` a template argument and use static_assert.

File: torch_tvm/operators.h (hunk 13 above)
Comment (yinghai) [reply torch_tvm/operators.h]: Yeah.

File: torch_tvm/managed_tensors.h (hunk 11 above)
Comment (yinghai) [reply torch_tvm/managed_tensors.h]: can't it just be `ManagedTensors`?

File: torch_tvm/managed_tensors.h [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+using torch::jit::Value;
+
+using torch_tvm::utils::DLManagedTensorPtr;
+
+class ManagedTensors {
+  public:
+    ManagedTensors() = default;
+    bool insert_value(const Value* val) {
```
Comment (yinghai) [inline torch_tvm/managed_tensors.h]: BTW, generally we use `insertValue` kind of naming convention?

File: torch_tvm/compiler.cpp [hunk 18]
```diff
@@ -320,8 +342,22 @@ void TVMCompiler::run(Stack& stack) {
...
     auto ivalue = value_to_ivalue.at(cache_[spec].input_values[i]);
-    auto tensor = ivalue.toTensor().to(at::kFloat);
-    auto dl_tensor = at::toDLPack(tensor);
+    //auto tensor = ivalue.toTensor().to(at::kFloat);
+    auto tensor = ivalue.toTensor();
+    DLManagedTensor* dl_tensor;
+    if (torch_tvm::utils::is_aligned(tensor.data_ptr(),
+          tvm::runtime::kAllocAlignment)) {
+      dl_tensor = at::toDLPack(tensor);
+    } else if (cache_[spec].tvm_param_tensors_.value_exists(value)) {
```
Comment (yinghai) [inline torch_tvm/compiler.cpp]: Eventually we don't want to even compare with this value if we already know what's treated as weights in TVM as you probably will go though alter layout and have your own transformed weights.

Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: yes, but we cant couple this with whether weights get transformed. Transformation is optimization step and not always done or can be relied upon. Right? Unless I am missing something.

File: torch_tvm/managed_tensors.h (hunk 17 above)
Comment (kimishpatel) [reply torch_tvm/managed_tensors.h]: Yes, removing this whole thing anyway.

File: torch_tvm/managed_tensors.h (hunk 11 above)
Comment (kimishpatel) [reply torch_tvm/managed_tensors.h]: Yes, I will remove this and do it alternatively.

File: torch_tvm/memory_utils.cpp (hunk 4 above)
Comment (kimishpatel) [reply torch_tvm/memory_utils.cpp]: Only if kAllocAlignment value which comes from tvm is constexpr. I will check and see.

File: torch_tvm/compiler.cpp (hunk 18 above)
Comment (yinghai) [reply torch_tvm/compiler.cpp]: Once you enable opt_lvl=3, you cannot expect that your weights are not changed.

Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: I dont think I am following you. The only expectation above condition sets is that we created and copied tensor that corresponds to some Value*. Here that tensor is weight. What happens inside tvm, as a result opt_lvl, is not what this intends to do.
Can you elaborate what you mean by "you cannot expect that your weights are not changed"? Are you talking about layout transformation of weights?

Comment (yinghai) [reply torch_tvm/compiler.cpp]: > Are you talking about layout transformation of weights?

Yes.

**Comment by kimishpatel** [comment]
@yinghai, thats a good point. For that we will have to make the aligned DLTensor that we created part of the `cache_` that caches compilation, so that DLTensor is destructed when that is destructed. In another PR.

File: torch_tvm/compiler.cpp (hunk 18 above)
Comment (kimishpatel) [reply torch_tvm/compiler.cpp]: Only reason the check is done is because: at the time of relay graph creation, the only thing we have is Value*. But this is not strictly true. We create relay graph on the very first inference which means the weights are definitely available. I think we can have `std::pair<Value*, DLManagedTensorPtr>`, instead of `std::pair<Value*, bool>` that maps Value* to its mutability property. But it probably makes code little less readable. Thanks for the input.
